File: ruby/medium/increasing_triplet_subsequence.rb
# @param {Integer[]} nums
# @return {Boolean}
def increasing_triplet(nums)
    return false if nums.length < 3
    i, j = 0, 1
    min_i, min_j = ((2 ** 31) - 1), ((2 ** 31) - 1)
    while j < nums.length
        return true if nums[j] > min_j
        if nums[j] <= nums[i]
            i = j
        else
            min_i = [min_i, nums[i]].min
            min_j = [min_j, nums[j]].min
        end
        j += 1
    end
    return false
end

File: javascript/medium/increasingTripletSubsequence.js
/**
 * @param {number[]} nums
 * @return {boolean}
 */
var increasingTriplet = function(nums) {
    if(nums.length < 3) return false;
    let i = 0, j = 1, min_i = Number.MAX_SAFE_INTEGER, min_j = Number.MAX_SAFE_INTEGER;
    while(j < nums.length) {
        if(nums[j] > min_j) return true;
        if(nums[j] > nums[i]) {
            min_i = Math.min(min_i, nums[i]);
            min_j = Math.min(min_j, nums[j]);
        } else {
            i = j;
        }
        j++;
    }
    return false;
};
